test/e2e/throughput_test.go

```diff
@@ -0,0 +1,201 @@
+		test.DeploymentScaledToZeroFunc,
+		"DeploymentScaledToZero",
+		test.ServingNamespace,
+		2*time.Minute); err != nil {
+		fmt.Errorf("Failed waiting for deployment to scale to zero: %v", err)
+	}
+
+	// Hit the service with concurrent requests to provoke the start from 0,
+	// the requests should go through activator, we make sure we receive only 200 responses.
+
+	logger.Info("Waiting for endpoint to serve request")
+
+	endpoint, err := spoof.GetServiceEndpoint(clients.KubeClient.Kube)
+	url := fmt.Sprintf("http://%s/?timeout=100", *endpoint)
+
+	client, err := pkgTest.NewSpoofingClient(clients.KubeClient, logger, domain, test.ServingFlags.ResolvableDomain)
+
+	totalRequests := 1000
+	responseChannel := make(chan *spoof.Response, totalRequests)
+	timeout := 30 * time.Second
+	roundTrip := roundTrip(client, url)
+
+	sendRequests(roundTrip, 1000, totalRequests, responseChannel, timeout, logger, t)
+
+	collectResponses(responseChannel, totalRequests, timeout, t)
+}
+
+func sendRequests(roundTrip func() (*spoof.Response, error), concurrency, totalRequests int, resChannel chan *spoof.Response, timeout time.Duration, logger *logging.BaseLogger, t *testing.T) int32 {
+	var (
+		group     errgroup.Group
+		responses int32
+	)
+	// The eventual number of sent requests is mod(concurrency).
+	runs := totalRequests / concurrency
+	tickChan := make(chan struct{}, runs)
+	doneChan := make(chan struct{})
+	timeoutChan := time.After(timeout)
+	errChan := make(chan error)
+	allDoneChan := make(chan struct{})
+
+	// Send a tick for each batch of requests.
+	go func() {
+		for i := 0; i < runs; i++ {
+			tickChan <- struct{}{}
+			<-doneChan
+		}
+		allDoneChan <- struct{}{}
+	}()
+
+	// Run a batch of requests.
+	go func() {
+		for {
+			select {
+			case <-tickChan:
+				logger.Info("Starting to execute the batch of requests")
+				//	Send requests async and wait for the responses.
+				for i := 0; i < concurrency; i++ {
+					group.Go(func() error {
+						res, err := roundTrip()
+						if err != nil {
+							errChan <- fmt.Errorf("Unexpected error sending a request, %v\n", err)
+						}
+						atomic.AddInt32(&responses, 1)
+						resChannel <- res
+						return nil
+					})
+				}
+				err := group.Wait()
+				if err != nil {
+					errChan <- fmt.Errorf("Unexpected error making requests against activator: %v", err)
+				}
+				logger.Info("Finished the execution of the requests")
+				doneChan <- struct{}{}
+				continue
+			case <-timeoutChan:
+				errChan <- fmt.Errorf("Timeout out waiting for responses, want - %d, got - %d", totalRequests, atomic.LoadInt32(&responses))
+				return
+			}
+		}
+	}()
+	logger.Info("Waiting for all requests to finish")
+	select {
+	case <-allDoneChan:
+		// success
+	case err := <-errChan:
+		t.Fatalf("Error happened while waiting for the responses: %v", err)
+	case <-timeoutChan:
+		t.Fatalf("Timeout waiting for resopnses")
+	}
+
+	logger.Info("All requests are finished")
+	return atomic.LoadInt32(&responses)
```

Comment:
you're no longer racing here, right? So just `return responses` should work?

Reply:
Race condition detector would complain I think.

Reply:
Well, if all the go routines terminate it shouldn't. You can check in as is, I can run experiment later myselc.